MinecraftForge update checker does not consider HTTP redirects.

```
[16:35:40] [Forge Version Check/DEBUG] [forge.VersionCheck]: [mousetweaks] Received version check data:
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.6.2</center>
</body>
</html>

[16:35:40] [Forge Version Check/DEBUG] [forge.VersionCheck]: Failed to process update information
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
	at com.google.gson.Gson.fromJson(Gson.java:899) ~[Gson.class:?]
	at com.google.gson.Gson.fromJson(Gson.java:852) ~[Gson.class:?]
	at com.google.gson.Gson.fromJson(Gson.java:801) ~[Gson.class:?]
	at com.google.gson.Gson.fromJson(Gson.java:773) ~[Gson.class:?]
	at net.minecraftforge.common.ForgeVersion$1.process(ForgeVersion.java:223) [ForgeVersion$1.class:?]
	at net.minecraftforge.common.ForgeVersion$1.run(ForgeVersion.java:198) [ForgeVersion$1.class:?]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:385) ~[JsonReader.class:?]
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:183) ~[MapTypeAdapterFactory$Adapter.class:?]
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145) ~[MapTypeAdapterFactory$Adapter.class:?]
	at com.google.gson.Gson.fromJson(Gson.java:887) ~[Gson.class:?]
	... 5 more
```

This is indicative that the remote server sent a 301 Moved Permanently, which, per the [relevant RFC](https://tools.ietf.org/html/rfc7231#section-6.4):

>    The 3xx (Redirection) class of status code indicates that further
>    action needs to be taken by the user agent in order to fulfill the
>    request.  If a Location header field (Section 7.1.2) is provided, the
>    user agent MAY automatically redirect its request to the URI
>    referenced by the Location field value, even if the specific status
>    code is not understood.  Automatic redirection needs to done with
>    care for methods not known to be safe, as defined in Section 4.2.1,
>    since the user might not wish to redirect an unsafe request.

--- a/src/main/java/net/minecraftforge/common/ForgeVersion.java
+++ b/src/main/java/net/minecraftforge/common/ForgeVersion.java
@@ -73,6 +73,8 @@ public class ForgeVersion
 
     private static final Logger log = LogManager.getLogger(MOD_ID + ".VersionCheck");
 
+    private static final int MAX_HTTP_REDIRECTS = Integer.getInteger("http.maxRedirects", 20);
+
     public static int getMajorVersion()
     {
         return majorVersion;
@@ -234,7 +236,7 @@ public class ForgeVersion
 
                     if (redirected)
                     {
-                        if (++redirects == 5)
+                        if (++redirects >= MAX_HTTP_REDIRECTS)
                         {
                             throw new IOException("Too many redirects while trying to fetch " + originalUrl);
                         }

--- a/src/main/java/net/minecraftforge/common/ForgeVersion.java
+++ b/src/main/java/net/minecraftforge/common/ForgeVersion.java
@@ -21,8 +21,11 @@ package net.minecraftforge.common;
 
 import static net.minecraftforge.common.ForgeVersion.Status.*;
 
+import java.io.IOException;
 import java.io.InputStream;
+import java.net.HttpURLConnection;
 import java.net.URL;
+import java.net.URLConnection;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
@@ -204,6 +207,46 @@ public class ForgeVersion
                 }
             }
 
+            /**
+             * Opens stream for given URL while following redirects (max. 5)
+             */
+            private InputStream openUrlStream(URL url) throws IOException
+            {
+                URL originalUrl = url;
+                int redirects = 0;
+                do
+                {
+                    URLConnection c = url.openConnection();
+                    boolean redirected = false;
+
+                    if (c instanceof HttpURLConnection)
+                    {
+                        HttpURLConnection huc = (HttpURLConnection) c;
+                        int responseCode = huc.getResponseCode();
+                        if (responseCode >= 301 && responseCode <= 303 || responseCode == 307 || responseCode == 308)
+                        {
+                            String loc = huc.getHeaderField("Location");
+                            url = new URL(loc);
+                            redirected = true;
+                            huc.disconnect();
+                        }
+                    }
+
+                    if (redirected)
+                    {
+                        if (++redirects == 5)
+                        {
+                            throw new IOException("Too many redirects while trying to fetch " + originalUrl);
+                        }
+                    }
+                    else
+                    {
+                        return c.getInputStream();
+                    }
+                }
+                while (true);
+            }
+
             private void process(ModContainer mod, URL url)
             {
                 try
@@ -212,7 +255,7 @@ public class ForgeVersion
                     Status status = PENDING;
                     ComparableVersion target = null;
 
-                    InputStream con = url.openStream();
+                    InputStream con = openUrlStream(url);
                     String data = new String(ByteStreams.toByteArray(con), "UTF-8");
                     con.close();
 

--- a/src/main/java/net/minecraftforge/common/ForgeVersion.java
+++ b/src/main/java/net/minecraftforge/common/ForgeVersion.java
@@ -228,7 +228,7 @@ public class ForgeVersion
                         if (responseCode >= 301 && responseCode <= 303 || responseCode == 307 || responseCode == 308)
                         {
                             String loc = huc.getHeaderField("Location");
-                            url = new URL(loc);
+                            url = new URL(url, loc);
                             redirected = true;
                             huc.disconnect();
                         }

--- a/src/main/java/net/minecraftforge/common/ForgeVersion.java
+++ b/src/main/java/net/minecraftforge/common/ForgeVersion.java
@@ -210,43 +210,37 @@ public class ForgeVersion
             }
 
             /**
-             * Opens stream for given URL while following redirects (max. 5)
+             * Opens stream for given URL while following redirects
              */
             private InputStream openUrlStream(URL url) throws IOException
             {
-                URL originalUrl = url;
-                int redirects = 0;
-                do
+                URL currentUrl = url;
+                for (int redirects = 0; redirects < MAX_HTTP_REDIRECTS; redirects++)
                 {
-                    URLConnection c = url.openConnection();
-                    boolean redirected = false;
-
+                    URLConnection c = currentUrl.openConnection();
                     if (c instanceof HttpURLConnection)
                     {
                         HttpURLConnection huc = (HttpURLConnection) c;
+                        huc.setInstanceFollowRedirects(false);
                         int responseCode = huc.getResponseCode();
-                        if (responseCode >= 301 && responseCode <= 303 || responseCode == 307 || responseCode == 308)
+                        if (responseCode >= 300 && responseCode <= 399)
                         {
-                            String loc = huc.getHeaderField("Location");
-                            url = new URL(url, loc);
-                            redirected = true;
-                            huc.disconnect();
+                            try
+                            {
+                                String loc = huc.getHeaderField("Location");
+                                currentUrl = new URL(currentUrl, loc);
+                                continue;
+                            }
+                            finally
+                            {
+                                huc.disconnect();
+                            }
                         }
                     }
 
-                    if (redirected)
-                    {
-                        if (++redirects >= MAX_HTTP_REDIRECTS)
-                        {
-                            throw new IOException("Too many redirects while trying to fetch " + originalUrl);
-                        }
-                    }
-                    else
-                    {
-                        return c.getInputStream();
-                    }
+                    return c.getInputStream();
                 }
-                while (true);
+                throw new IOException("Too many redirects while trying to fetch " + url);
             }
 
             private void process(ModContainer mod, URL url)

